signal/src/lib.rs

```diff
@@ -97,914 +116,656 @@ pub mod pallet {
-
-				// TODO:
-				// a. result( accepted, rejected )
-				// b. result( accepted, rejected, total_allowed )
-				// c. result( required_majority, staked_accept, staked_reject, slash_amount )
-				// d. threshold reached
-				// e. conviction
-
-				match &proposal.proposal_type {
-					ProposalType::General => {
-						// simple vote
-						let (yes, no) = ProposalSimpleVotes::<T>::get(&proposal_id);
-						if yes > no {
-							proposal_state = ProposalState::Accepted;
-						}
-						if yes <= no {
-							proposal_state = ProposalState::Rejected;
+				let voting =  ProposalVoting::<T>::get(&proposal_id).unwrap();
+
+				// Get the final state based on Voting participation, quorum, majority
+				proposal_state = Self::get_final_proposal_state(&voting);
+				
+				Self::finalize_proposal(&proposal_id, proposal_state, &voting);
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+
+		pub fn get_voting_power(voting: &Voting<T>, deposit: &Option<T::Balance>) -> VotingPower {
+			let mut power: VotingPower = 1;
+			match voting.unit {
+				Unit::Person => {
+					match voting.scale {
+						Scale::Linear => { 
+							power = 1;
 						}
-						if yes == 0 && no == 0 {
-							proposal_state = ProposalState::Expired;
+						Scale::Quadratic => { 
+							// So far not possible, maybe in case of delegation
 						}
 					}
-					ProposalType::Withdrawal => {
-						// treasury
-						// 50% majority of eligible voters
-						let (yes, _no) = ProposalSimpleVotes::<T>::get(&proposal_id);
-						let campaign_id = &proposal.campaign_id.unwrap();
-						let contributors = T::Flow::campaign_contributors_count(&campaign_id);
-						// TODO: dynamic threshold
-						let threshold = contributors.checked_div(2).ok_or(Error::<T>::DivisionError);
-						match threshold {
-							Ok(t) => {
-								if yes > t {
-									proposal_state = ProposalState::Accepted;
-									// TODO: handle an error
-									let res = Self::unlock_balance(&proposal, yes);
-									debug_assert!(res.is_ok());
-								} else {
-									proposal_state = ProposalState::Rejected;
-								}
-							}
-							Err(_err) => {
-								// todo: logic on error event
-							}
+				}
+				Unit::Token => {
+					let linear_power: VotingPower = deposit.unwrap().saturated_into();
+					match voting.scale {
+						Scale::Linear => { 
+							power = linear_power;
+						}
+						Scale::Quadratic => { 
+							let linear_power: VotingPower = deposit.unwrap().saturated_into();
+							power = linear_power.integer_sqrt();
 						}
-					}
-					ProposalType::Member => {
-						// membership
-						//
-					}
-					_ => {
-						// no result - fail
-						proposal_state = ProposalState::Expired;
 					}
 				}
+			}
+			power
+		}
 
-				ProposalStates::<T>::insert(&proposal_id, proposal_state.clone());
+		pub fn process_voting_deposits(
+			who: &T::AccountId,
+			old_deposit: &Option<T::Balance>,
+			deposit: &Option<T::Balance>
+		) -> Result<(), DispatchError> {
```

Comment:
Here, if `old_deposit == deposit`, there is no need to perform any action.
But instead, it will create two additional operations.